p11-kit/rpc-message.c

```diff
@@ -1660,6 +1660,72 @@ p11_rpc_buffer_get_aes_ctr_mechanism_value (p11_buffer *buffer,
 	return true;
 }
 
+void
+p11_rpc_buffer_add_aes_gcm_mechanism_value (p11_buffer *buffer,
+					    const void *value,
+					    CK_ULONG value_length)
+{
+	CK_GCM_PARAMS params;
+
+	/* Check if value can be converted to CK_GCM_PARAMS. */
+	if (value_length != sizeof (CK_GCM_PARAMS)) {
+		p11_buffer_fail (buffer);
+		return;
+	}
+
+	memcpy (&params, value, value_length);
+
+	/* Check if params.ulTagBits can be converted to uint64_t. */
+	if (params.ulTagBits > UINT64_MAX) {
+		p11_buffer_fail (buffer);
+		return;
+	}
+
+	p11_rpc_buffer_add_byte_array (buffer,
+				       (unsigned char *)params.pIv,
+				       params.ulIvLen);
+	p11_rpc_buffer_add_byte_array (buffer,
+				       (unsigned char *)params.pAAD,
+				       params.ulAADLen);
+	p11_rpc_buffer_add_uint64 (buffer, params.ulTagBits);
+}
+
+bool
+p11_rpc_buffer_get_aes_gcm_mechanism_value (p11_buffer *buffer,
+					    size_t *offset,
+					    void *value,
+					    CK_ULONG *value_length)
+{
+	uint64_t val;
+	const unsigned char *data1, *data2;
+	size_t len1, len2;
+
+	if (!p11_rpc_buffer_get_byte_array (buffer, offset, &data1, &len1))
+		return false;
+	if (!p11_rpc_buffer_get_byte_array (buffer, offset, &data2, &len2))
+		return false;
+	if (!p11_rpc_buffer_get_uint64 (buffer, offset, &val))
+		return false;
+
+	if (value) {
+		CK_GCM_PARAMS params;
+
+		params.pIv = (void *) data1;
+		params.ulIvLen = len1;
+		params.ulIvBits = len1 * 8;
```

Comment:
Maybe I should better transport the `ulIvBits` field separately and not derive it from `ulIvLen`? I guess it does not necessarily correspond directly to `ulIvLen * 8`, but can be less.

What do you think?

Reply:
According to PKCS#11 3.0, `ulIvLen` seems to take precedence over `ulIvBits` so I guess we could keep it.
By the way I wonder if we can handle AES-GCM properly with the PKCS#11 2.40 interface alone, which lacks message based [functions](https://www.oasis-open.org/committees/download.php/52895/PKCS11MessageBasedEncryption20140502.pdf).

Reply:
So do we want to keep it as is, or do we want to transport ulIvBits separately? 

This would be something like the following:

```
 p11-kit/rpc-message.c | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

diff --git a/p11-kit/rpc-message.c b/p11-kit/rpc-message.c
index 97b5a30..f785b79 100644
--- a/p11-kit/rpc-message.c
+++ b/p11-kit/rpc-message.c
@@ -1675,8 +1675,8 @@ p11_rpc_buffer_add_aes_gcm_mechanism_value (p11_buffer *buffer,
 
 	memcpy (&params, value, value_length);
 
-	/* Check if params.ulTagBits can be converted to uint64_t. */
-	if (params.ulTagBits > UINT64_MAX) {
+	/* Check if params.ulTagBits/ulIvBits can be converted to uint64_t. */
+	if (params.ulTagBits > UINT64_MAX || params.ulIvBits > UINT64_MAX) {
 		p11_buffer_fail (buffer);
 		return;
 	}
@@ -1684,6 +1684,7 @@ p11_rpc_buffer_add_aes_gcm_mechanism_value (p11_buffer *buffer,
 	p11_rpc_buffer_add_byte_array (buffer,
 				       (unsigned char *)params.pIv,
 				       params.ulIvLen);
+	p11_rpc_buffer_add_uint64 (buffer, params.ulIvBits);
 	p11_rpc_buffer_add_byte_array (buffer,
 				       (unsigned char *)params.pAAD,
 				       params.ulAADLen);
@@ -1696,15 +1697,17 @@ p11_rpc_buffer_get_aes_gcm_mechanism_value (p11_buffer *buffer,
 					    void *value,
 					    CK_ULONG *value_length)
 {
-	uint64_t val;
+	uint64_t val1, val2;
 	const unsigned char *data1, *data2;
 	size_t len1, len2;
 
 	if (!p11_rpc_buffer_get_byte_array (buffer, offset, &data1, &len1))
 		return false;
+	if (!p11_rpc_buffer_get_uint64 (buffer, offset, &val1))
+		return false;
 	if (!p11_rpc_buffer_get_byte_array (buffer, offset, &data2, &len2))
 		return false;
-	if (!p11_rpc_buffer_get_uint64 (buffer, offset, &val))
+	if (!p11_rpc_buffer_get_uint64 (buffer, offset, &val2))
 		return false;
 
 	if (value) {
@@ -1712,10 +1715,10 @@ p11_rpc_buffer_get_aes_gcm_mechanism_value (p11_buffer *buffer,
 
 		params.pIv = (void *) data1;
 		params.ulIvLen = len1;
-		params.ulIvBits = len1 * 8;
+		params.ulIvBits = val1;
 		params.pAAD = (void *) data2;
 		params.ulAADLen = len2;
-		params.ulTagBits = val;
+		params.ulTagBits = val2;
 
 		memcpy (value, &params, sizeof (CK_GCM_PARAMS));
 	}
```

Let me know what you prefer.

Reply:
Thanks; I like this change.

Reply:
changed.